File: Algorithms/Maths/Java/Karatsuba_Algorithm.java
import java.util.Scanner;

public class Karatsuba_Algorithm {
    public static long Karatsuba(long x, long y){
        if(x < 10 || y < 10){
            return (x * y);
        }

        int n = Math.max(Long.toString(x).length(), Long.toString(y).length());
        int half = (n + 1) / 2;
        
        long a = x / (long) Math.pow(10, half);
        long b = x % (long) Math.pow(10, half);
        long c = x / (long) Math.pow(10, half);
        long d = x % (long) Math.pow(10, half);

        long ac = Karatsuba(a, c);
        long bd = Karatsuba(b, d);
        long adbc = Karatsuba(a + b, c + d) - ac - bd;

        return (long) (ac * Math.pow(10, 2 * half) + adbc * Math.pow(10, half) + bd);
    }
    public static void main(String[] args){
        
        Scanner sc = new Scanner(System.in);

        System.out.print("Enter the First Number: ");
        long n1 = sc.nextLong();
        System.out.print("Enter the Second Number: ");
        long n2 = sc.nextLong();

        System.out.println("The Multiplication of " + n1 + " and " + n2 + " is: " + Karatsuba(n1, n2));
        
        sc.close();
    }
}

File: Algorithms/Maths/C/Karatsuba_Algorithm.c
#include<stdio.h>
#include<math.h>
#define MAX(i, j) (((i) > (j)) ? (i) : (j))

long Karatsuba(long x, long y){
    if(x < 10 || y < 10){
        return (x * y);
    }

    int n, half, count_x = 0, count_y = 0, x_copy = x, y_copy = y;
    long a, b, c, d, ac, bd, adbc;

    while (x_copy != 0){
        x_copy /= 10;
        ++count_x;
    }

    while (y_copy != 0){
        y_copy /= 10;
        ++count_y;
    }

    n = MAX(count_x, count_y);
    half = (n + 1) / 2;

    a = x / (long) pow(10.0, (double)half);
    b = x % (long) pow(10.0, (double)half);
    c = y / (long) pow(10.0, (double)half);
    d = y % (long) pow(10.0, (double)half);

    ac = Karatsuba(a, c);
    bd = Karatsuba(b, d);
    adbc = Karatsuba(a + b, c + d) - ac - bd;

    return (long) (ac * pow(10.0, (double) (2 * half)) + adbc * pow(10.0, (double)half) + bd);
}

int main(){

    long n1, n2;
    printf("Enter the First Number: ");
    scanf("%ld", &n1);
    printf("Enter the Second Number: ");
    scanf("%ld", &n2);

    printf("The Multiplication of %ld and %ld is: %ld", n1, n2, Karatsuba(n1, n2));

    return 0;
}